### PR TITLE
remove liblzma hacks in ffmpeg and mpv

### PR DESCRIPTION
**Pull request by tomty89**
let's see if runpath actually works

### CODE REVIEW

**Comment by xeffyr** [comment]
Works without linking errors.

**Comment by xeffyr** [comment]
Will see if liblzma with `--enable-symbol-versions` changes something.

**Comment by xeffyr** [comment]
Versioned symbols does not solve problem.

But if `mpv` is executed in way like this:
```
LD_PRELOAD=/system/lib64/liblzma.so:$PREFIX/lib/liblzma.so mpv
```
it will work.

**Comment by tomty89** [comment]
Have you rebuilt both ffmpeg and mpv after enabling versioned symbol in liblzma? (So that the ffmpeg libraries and mpv has versioned linkage as shown in `ldd` or `readelf --dynamic`.)

If so, what exactly is the error?

**Comment by xeffyr** [comment]
Yes, ffmpeg, mpv, libarchive were rebuilt.

mpv's error:
```
$ mpv
CANNOT LINK EXECUTABLE "mpv": cannot locate symbol "XzUnpacker_Construct" referenced by "/system/lib64/libunwind.so"...
Aborted
```

liblzma has versioned symbols (`objdump -T $PREFIX/lib/liblzma.so`):
```
***
0000000000009e68 g    DF .text  000000000000000c  XZ_5.0      lzma_index_hash_size
000000000000a314 g    DF .text  0000000000000190  XZ_5.0      lzma_stream_buffer_decode
000000000000aa8c g    DF .text  0000000000000070  XZ_5.0      lzma_stream_decoder
000000000000abf0 g    DF .text  00000000000000b0  XZ_5.0      lzma_stream_footer_decode
000000000000ab14 g    DF .text  000000000000009c  XZ_5.0      lzma_stream_header_decode
000000000000b020 g    DF .text  00000000000000e4  XZ_5.0      lzma_crc64
000000000000c7d8 g    DF .text  000000000000002c  XZ_5.0      lzma_mf_is_supported
000000000000e910 g    DF .text  0000000000000010  XZ_5.0      lzma_mode_is_supported
0000000000000000 g    DO *ABS*  0000000000000000  XZ_5.0      XZ_5.0
0000000000000000 g    DO *ABS*  0000000000000000  XZ_5.2      XZ_5.2
```

Full list of symbols for mpv, libarchive, liblzma and libavcodec: [logs.tar.gz](https://github.com/termux/termux-packages/files/3141480/logs.tar.gz)

System liblzma is unversioned and I guess this is a problem. It also entirely different implementation sources for which available at https://android.googlesource.com/platform/external/lzma/. I'm even thinking of moving `XzUnpacker_Construct()` and some others to Termux's liblzma to see if it become compatible.

**Comment by xeffyr** [comment]
Just for experiment I have modified liblzma.so in way that it now contains functions from Android's /system/lib64/liblzma.so:
```
$ objdump -T liblzma.so | grep XzUnpacker
000000000001432c g    DF .text  0000000000000058  Base        XzUnpacker_IsBlockFinished
00000000000143e8 g    DF .text  0000000000000074  Base        XzUnpacker_GetExtraSize
0000000000012018 g    DF .text  0000000000001334  Base        XzUnpacker_Code
0000000000011cec g    DF .text  0000000000000060  Base        XzUnpacker_Init
0000000000011e74 g    DF .text  000000000000002c  Base        XzUnpacker_Free
0000000000014384 g    DF .text  0000000000000064  Base        XzUnpacker_IsStreamWasFinished
0000000000011d7c g    DF .text  0000000000000058  Base        XzUnpacker_Construct
0000000000011d4c g    DF .text  0000000000000030  Base        XzUnpacker_SetOutBuf
00000000000142a8 g    DF .text  0000000000000084  Base        XzUnpacker_CodeFull
0000000000011fac g    DF .text  000000000000006c  Base        XzUnpacker_PrepareToRandomBlockDecoding
```
Library download (aarch64): [liblzma.tar.gz](https://github.com/termux/termux-packages/files/3141724/liblzma.tar.gz)


If I copy it to prefix, everything will work:
```
[~]:$ mpv
CANNOT LINK EXECUTABLE "mpv": cannot locate symbol "XzUnpacker_Construct" referenced by "/system/lib64/libunwind.so"...
Aborted
[~]:$ cp liblzma.so $PREFIX/lib/
cp: overwrite '/data/data/com.termux/files/usr/lib/liblzma.so'? y
[~]:$ mpv --version
mpv 0.29.1 Copyright © 2000-2018 mpv/MPlayer/mplayer2 projects
 built on Thu May  2 16:26:12 UTC 2019
ffmpeg library versions:
   libavutil       56.22.100
   libavcodec      58.35.100
   libavformat     58.20.100
   libswscale      5.3.100
   libavfilter     7.40.101
   libswresample   3.3.100
ffmpeg version: 4.1.3
```

**Comment by tomty89** [comment]
That's not what I meant. I meant have you confirmed that the `SONAME` of our liblzma is versioned (and hence the `NEEDED` of it in mpv and ffmpeg libraries).

**Comment by xeffyr** [comment]
liblzma.so:
```
 0x000000000000000e (SONAME)             Library soname: [liblzma.so]
```

mpv:
```
 0x0000000000000004 (HASH)               0x6380
 0x0000000000000001 (NEEDED)             Shared library: [libavcodec.so]
 0x0000000000000001 (NEEDED)             Shared library: [libpulse.so]
 0x0000000000000001 (NEEDED)             Shared library: [libOpenSLES.so]
 0x0000000000000001 (NEEDED)             Shared library: [libm.so]
 0x0000000000000001 (NEEDED)             Shared library: [liblua5.2.so.5.2]
 0x0000000000000001 (NEEDED)             Shared library: [libopenal.so]
 0x0000000000000001 (NEEDED)             Shared library: [libavutil.so]
 0x0000000000000001 (NEEDED)             Shared library: [libavformat.so]
 0x0000000000000001 (NEEDED)             Shared library: [libswscale.so]
 0x0000000000000001 (NEEDED)             Shared library: [libavfilter.so]
 0x0000000000000001 (NEEDED)             Shared library: [libswresample.so]
 0x0000000000000001 (NEEDED)             Shared library: [libz.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libarchive.so]
 0x0000000000000001 (NEEDED)             Shared library: [libandroid-support.so]
 0x0000000000000001 (NEEDED)             Shared library: [libandroid-glob.so]
 0x0000000000000001 (NEEDED)             Shared library: [libdl.so]
 0x0000000000000001 (NEEDED)             Shared library: [libc.so]
 0x0000000000000020 (PREINIT_ARRAY)      0x13a9e0
```

**Comment by xeffyr** [comment]
mpv links with libarchive but not with liblzma.

**Comment by xeffyr** [comment]
libavcodec:
```
 0x0000000000000001 (NEEDED)             Shared library: [liblzma.so]
```

**Comment by xeffyr** [comment]
Found in configure script:
```
linux*android*)
  version_type=none # Android doesn't support versioned libraries.
  need_lib_prefix=no
  need_version=no
```

**Comment by xeffyr** [comment]
Versioned liblzma works.

**Comment by tomty89** [comment]
Ahh right, it was libarchive that needed to be rebuilt after liblzma is versioned.

File: packages/liblzma/build.sh [hunk 1]
```diff
@@ -2,5 +2,7 @@ TERMUX_PKG_HOMEPAGE=https://tukaani.org/xz/
 TERMUX_PKG_DESCRIPTION="XZ-format compression library"
 TERMUX_PKG_LICENSE="GPL-2.0"
 TERMUX_PKG_VERSION=5.2.4
+TERMUX_PKG_REVISION=1
 TERMUX_PKG_SHA256=9717ae363760dedf573dad241420c5fea86256b65bc21d2cf71b2b12f0544f4b
 TERMUX_PKG_SRCURL=https://fossies.org/linux/misc/xz-${TERMUX_PKG_VERSION}.tar.xz
+TERMUX_PKG_EXTRA_CONFIGURE_ARGS="--enable-symbol-versions"
```
Comment (tomty89) [inline packages/liblzma/build.sh]: Not useful/necessary?

Comment (xeffyr) [reply packages/liblzma/build.sh]: Looks like not necessary as system's liblzma has different function names.

**Pull request merged**